openhands/runtime/base.py

```diff
@@ -131,15 +159,28 @@ async def on_event(self, event: Event) -> None:
             if event.timeout is None:
                 event.timeout = self.config.sandbox.timeout
             assert event.timeout is not None
-            observation: Observation = await call_sync_from_async(
-                self.run_action, event
-            )
+            try:
+                observation: Observation = await call_sync_from_async(
+                    self.run_action, event
+                )
+            except Exception as e:
+                err_id = ''
+                if isinstance(e, ConnectionError) or isinstance(
+                    e, RuntimeDisconnectedError
+                ):
+                    err_id = 'STATUS$ERROR_RUNTIME_DISCONNECTED'
+                self.log('error', f'Unexpected error while running action {e}')
+                self.log('error', f'Problematic action: {str(event)}')
+                self.send_error_message(err_id, str(e))
+                self.close()
+                return
```

Comment:
Just a quick question here. This doesn't has no obs, and doesn't re-raise (shouldn't it?), while the loop is waiting for an obs. I tried to replicate this manually somehow and I raised just in the `try` block:
- running with `main.py`
- it close()ed runtime
- then it looks like it hangs and that's it. Loop forever.

Reply:
Hmm this is a good point. Maybe it should reraise. Will try

Reply:
I don't think it works yet. First, this results in a "coro not awaited". Second, if the exception is not propagated downstream and doesn't stop the loop, then it would still have `_pending_action` set and I don't see why that will not loop endlessly...

Reply:
OK I refactored a bunch here, and now main.py and cli.py behave nicely in the presence of fatal exceptions. took some plumbing though--could definitely simplify further in a future PR